media-sound/freac/freac-1.1~alpha20180913.recipe

```diff
@@ -0,0 +1,93 @@
+SUMMARY="A free and open source audio converter"
+DESCRIPTION="fre:ac is a free and open source audio converter. It comes with \
+an integrated tag editor and converts between various formats.
+
+Features include:
+
+- Support for MP3, MP4/AAC, Opus, FLAC and other formats
+- Integrated player and tag editor
+- Support for playlists, cue sheets and chapters
+- Fast, multi-threaded conversion engine
+- Ability to convert to multiple formats at once
+- Can verify lossless input and output files
+- Support for freedb/CDDB queries and submits
+- Support for using command line codecs
+- User interface available in 40+ languages
+- Optional command line interface"
+HOMEPAGE="https://freac.org/"
+COPYRIGHT="2001-2018 Robert Kausch"
+LICENSE="GNU GPL v2"
+REVISION="1"
+SOURCE_URI="https://github.com/enzo1982/freac/releases/download/v${portVersion/\~alpha/-alpha-}/freac-${portVersion/\~alpha/-alpha-}.tar.gz"
+CHECKSUM_SHA256="5ca1f3b8f548815ced94e8df3b0b736d75879816939661ba1f1baa94fd36f168"
+SOURCE_DIR="freac-${portVersion/\~alpha/-alpha-}"
+PATCHES="freac-$portVersion.patchset"
+ADDITIONAL_FILES="freac.rdef.in"
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
+
+PROVIDES="
+	freac$secondaryArchSuffix = $portVersion
+	cmd:freac = $portVersion
+	cmd:freaccmd = $portVersion
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	cmd:ffmpeg
+	cmd:mpcdec
+	cmd:mpcenc
+	cmd:wavpack
+	cmd:wvunpack
+	lib:libsmooth_0.8.74$secondaryArchSuffix >= 0
+	lib:libboca_1.0$secondaryArchSuffix >= 0
+	lib:libcdio$secondaryArchSuffix
+	lib:libcdio_cdda$secondaryArchSuffix
+	lib:libcdio_paranoia$secondaryArchSuffix
+	lib:libfaac$secondaryArchSuffix
+	lib:libfaad$secondaryArchSuffix
+	lib:libFLAC$secondaryArchSuffix
+	lib:libmp3lame$secondaryArchSuffix
+	lib:libmp4v2$secondaryArchSuffix
+	lib:libmpg123$secondaryArchSuffix
+	lib:libogg$secondaryArchSuffix
+	lib:libopus$secondaryArchSuffix
+	lib:librubberband$secondaryArchSuffix
+	lib:libsamplerate$secondaryArchSuffix
+	lib:libsndfile$secondaryArchSuffix
+	lib:libspeex$secondaryArchSuffix
+	lib:libvorbis$secondaryArchSuffix
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libboca_1.0$secondaryArchSuffix >= 0
+	"
+BUILD_PREREQUIRES="
+	cmd:g++$secondaryArchSuffix
+	cmd:gcc$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	"
+
+BUILD()
+{
+	make $jobArgs
+}
+
+INSTALL()
+{
+	make prefix="$prefix" bindir="$prefix"/bin libdir="$libDir" includedir="$includeDir" datadir="$dataDir" install
```

Comment:
This will work on a secondary architecture, but will put `freac.so` in `lib/freac` instead of `lib:x86/freac`. fre:ac is searching for it in `lib/freac` which is why you got the error message yesterday and why it works with your changes.

BoCA would have to be adjusted to put its plugins in `lib/boca` too. Not sure if this is acceptable. I'd rather like to fix this in an upcoming release and re-add secondary architecture support then.

Reply:
So freac uses it's plugins from BoCA? It finds the boca library ok so far

Reply:
The BoCA library is linked in at link-time, so it will always be found as long as it is in the correct library path. `freac.so` and all the plugins (codecs, extensions etc.) are linked dynamically at runtime, though. fre:ac and BoCA look for these files in `lib/freac` and `lib/boca` which is why it doesn't work with the files in `lib:x86/freac` and `lib:x86/boca`.

The relevant lines in the code are [enzo1982/freac/blob/master/src/loader/gui.cpp#L34](https://github.com/enzo1982/freac/blob/master/src/loader/gui.cpp#L34) and [enzo1982/BoCA/blob/master/runtime/common/utilities.cpp#L67](https://github.com/enzo1982/BoCA/blob/master/runtime/common/utilities.cpp#L67). These will have to be changed in a new release to fix this.

Reply:
Ok, makes sense now :) In that case you could use !x86 (set it as broken) for now in the ARCHITECTURES instead of leaving the SECONDARY_ARCHITECURE out (once it's fixed the only change would be to use "x86" then instead of "!x86"?

Reply:
We can insert a `PATCH()` function between `BUILD_PREREQUIRES` and `BUILD`:
```
	cmd:pkg_config$secondaryArchSuffix
	"

PATCH()
{
	if [ -n "$secondaryArchSuffix" ]; then
		sed -i \
			-e "s|lib/freac|$relativeLibDir/freac|" \
			src/loader/gui.cpp
		sed -i \
			-e "s|lib/boca/|$relativeLibDir/boca|" \
			runtime/common/utilities.cpp
	fi
}

BUILD()
{
```
OK, this is a bit dirty..

Warning: I did not test this.

Reply:
Hmm, it might not work because this hack would probably require that fre!a, BoCA and smooth all have the same `$prefix`, but this is not true on Haiku.

Reply:
How about this:
```
	if [ -n "$secondaryArchSuffix" ]; then
		sed -i \
			-e "s|../lib/boca/|../lib:$secondaryArchSuffix/boca|" \
			runtime/common/utilities.cpp
	fi
```
What comes after the `:` in `lib:...` should be the same as the secondary arch suffix, no?

Reply:
OK, I ended up with this:
```
	if [ -n "$secondaryArchSuffix" ]; then
		sed -i \
			-e "s|/lib/freac|/lib/${secondaryArchSuffix#_}/freac|" \
			src/loader/gui.cpp
	fi
```
Likewise for BoCA and smooth and I now have a working fre:ac on Haiku x86_gcc2-hybrid. Will post the updated scripts tomorrow after some more testing.

Reply:
Although `/lib/${secondaryArchSuffix#_}/` will work for both primary and secondary arches, on primary arches it will be equivalent to `/lib//`, while if you use `/$relativeLibDir/` (instead of `/lib/${secondaryArchSuffix#_}/`) then it will be `/lib/` on any primary arch, and `/lib/x86/` on x86 secondary arch.
That's why I suggested using `-e "s|lib/freac|$relativeLibDir/freac|"` in https://github.com/haikuports/haikuports/pull/3197#discussion_r224011668.

Reply:
OK, I went with `$relativeLibDir`.